autopush-common/src/db/bigtable/bigtable_client/merge.rs

```diff
@@ -0,0 +1,486 @@
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::time::{Duration, SystemTime};
+
+use futures::StreamExt;
+use google_cloud_rust_raw::bigtable::v2::bigtable::{ReadRowsResponse, ReadRowsResponse_CellChunk};
+use grpcio::ClientSStreamReceiver;
+
+use super::{cell::Cell, error::BigTableError, row::Row, FamilyId, Qualifier, RowKey};
+
+/// List of the potential states when we are reading each value from the
+/// returned stream and composing a "row"
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+enum ReadState {
+    #[default]
+    RowStart,
+    CellStart,
+    CellInProgress,
+    CellComplete,
+    RowComplete,
+}
+
+/// An in-progress Cell data struct.
+#[derive(Debug, Clone)]
+pub(crate) struct PartialCell {
+    pub(crate) family: FamilyId,
+    /// A "qualifier" is a column name
+    pub(crate) qualifier: Qualifier,
+    /// Timestamps are returned as microseconds, but need to be
+    /// specified as milliseconds (even though the function asks
+    /// for microseconds, you * 1000 the mls).
+    pub(crate) timestamp: SystemTime,
+    /// Not sure if or how these are used
+    pub(crate) labels: Vec<String>,
+    /// The data buffer.
+    pub(crate) value: Vec<u8>,
+    /// the returned sort order for the cell.
+    pub(crate) value_index: usize,
+}
+
+impl Default for PartialCell {
+    fn default() -> Self {
+        Self {
+            family: String::default(),
+            qualifier: String::default(),
+            timestamp: SystemTime::now(),
+            labels: Vec::new(),
+            value: Vec::new(),
+            value_index: 0,
+        }
+    }
+}
+
+/// An in-progress Row structure
+#[derive(Debug, Default, Clone)]
+pub(crate) struct PartialRow {
+    /// table uniquie key
+    row_key: RowKey,
+    /// map of cells per family identifier.
+    cells: HashMap<FamilyId, Vec<PartialCell>>,
+    /// the last family id string we encountered
+    last_family: FamilyId,
+    /// the working set of family and cells
+    /// we've encountered so far
+    last_family_cells: HashMap<FamilyId, Vec<Cell>>,
+    /// the last column name we've encountered
+    last_qualifier: Qualifier,
+    /// Any cell that may be in progress (chunked
+    /// across multiple portions)
+    cell_in_progress: RefCell<PartialCell>,
```

Comment:
Is this `RefCell` necessary?

I was able to [kill the one on `PartialRow`](https://github.com/mozilla-services/autopush-rs/compare/feat/big_table...feat/big_table-partialrow-refcell) -- I think the key there was utilizing `std::mem::take` since the fields on these structs are simple types with defaults.

Reply:
that branch now kills both `RefCell`s